### PR TITLE
do not create the trafo

### PR DESCRIPTION
**Pull request by fieker**

### CODE REVIEW

File: src/flint/fmpz_mat.jl [hunk 1]
```diff
@@ -934,9 +934,8 @@ Given the Gram matrix $x$ of a matrix, compute the Gram matrix of its LLL
 reduction inplace.
 """
 function lll_gram!(x::fmpz_mat, ctx::lll_ctx = lll_ctx(0.99, 0.51, :gram))
-   u = similar(x, nrows(x), nrows(x))
    ccall((:fmpz_lll, libflint), Nothing,
-         (Ref{fmpz_mat}, Ref{fmpz_mat}, Ref{lll_ctx}), x, u, ctx)
+         (Ref{fmpz_mat}, Ptr{nothing}, Ref{lll_ctx}), x, C_NULL, ctx)
```
Comment (thofma) [inline src/flint/fmpz_mat.jl:938]:
```suggestion
         (Ref{fmpz_mat}, Ptr{Nothing}, Ref{lll_ctx}), x, C_NULL, ctx)
```

File: src/flint/fmpz_mat.jl [hunk 2]
```diff
@@ -968,9 +967,8 @@ are the rows remaining after removal.
 """
 function lll_with_removal(x::fmpz_mat, b::fmpz, ctx::lll_ctx = lll_ctx(0.99, 0.51))
    z = deepcopy(x)
-   u = similar(x, nrows(x), nrows(x))
    d = Int(ccall((:fmpz_lll_with_removal, libflint), Cint,
-    (Ref{fmpz_mat}, Ref{fmpz_mat}, Ref{fmpz}, Ref{lll_ctx}), z, u, b, ctx))
+    (Ref{fmpz_mat}, Ptr{nothing}, Ref{fmpz}, Ref{lll_ctx}), z, C_NULL, b, ctx))
```
Comment (thofma) [inline src/flint/fmpz_mat.jl:971]:
```suggestion
    (Ref{fmpz_mat}, Ptr{Nothing}, Ref{fmpz}, Ref{lll_ctx}), z, C_NULL, b, ctx))
```

File: src/flint/fmpz_mat.jl (hunk 1 above)
Comment (thofma) [reply src/flint/fmpz_mat.jl:938]: I think it is just a coincidence that it still works.

**Comment by fieker** [comment]
On Thu, Jan 07, 2021 at 03:38:39AM -0800, thofma wrote:
> @thofma commented on this pull request.
> 
> 
> 
> >     ccall((:fmpz_lll, libflint), Nothing,
> -         (Ref{fmpz_mat}, Ref{fmpz_mat}, Ref{lll_ctx}), x, u, ctx)
> +         (Ref{fmpz_mat}, Ptr{nothing}, Ref{lll_ctx}), x, C_NULL, ctx)
> 
> I think it is just a coincidence that it still works.

I think nothing is a legal julia objetc, hence can have a Ptr.
Experimentally: Ptr{bla}(C_NULL) wotks for all bla, be it objects or
types...
Is there a deprication looming?

**Comment by thofma** [comment]
No, but `typeof(C_NULL) === Ptr{Nothing}`. That something like `Ptr{nothing}` works is just a coincidence. According to the documentation `Ptr{T}` wants `T` to be a type, not a value.

**Comment by fieker** [comment]
On Thu, Jan 07, 2021 at 03:51:10AM -0800, thofma wrote:
> No, but `typeof(C_NULL) === Ptr{Nothing}`. That something like `Ptr{nothing}` works is just a coincidence. According to the documentation `Ptr{T}` wants `T` to be a type, not a value.
> 
Ptr{1} works....

**Comment by thofma** [comment]
As I said, it is a coincidence because those are bits types and they can be type parameters. But `Ptr{[1]}` does not work.

**Pull request merged**